types/snapsvg/index.d.ts

```diff
@@ -1,74 +1,109 @@
-    export function matrix(svgMatrix:SVGMatrix):Matrix;
-
-    export function ajax(url:string,postData:string,callback:Function,scope?:Object):XMLHttpRequest;
-    export function ajax(url:string,postData:Object,callback:Function,scope?:Object):XMLHttpRequest;
-    export function ajax(url:string,callback:Function,scope?:Object):XMLHttpRequest;
-    export function format(token:string,json:Object):string;
-    export function fragment(varargs:any):Fragment;
-    export function getElementByPoint(x:number,y:number):Snap.Element;
-    export function is(o:any,type:string):boolean;
-    export function load(url:string,callback:Function,scope?:Object):void;
-    export function plugin(f:Function):void;
-    export function select(query:string):Snap.Element;
-    export function selectAll(query:string):any;
-    export function snapTo(values:Array<number>,value:number,tolerance?:number):number;
-
-    export function animate(from:number|number[],to:number|number[],updater:(n:number)=>void,duration:number,easing?:(num:number)=>number,callback?:()=>void):mina.MinaAnimation;
-    export function animation(attr:Object,duration:number,easing?:(num:number)=>number,callback?:()=>void):Snap.Animation;
-
-    export function color(clr:string):RGBHSB;
-    export function getRGB(color:string):RGB;
-    export function hsb(h:number,s:number,b:number):HSB;
-    export function hsl(h:number,s:number,l:number):HSL;
-    export function rgb(r:number,g:number,b:number):RGB;
-    export function hsb2rgb(h:number,s:number,v:number):RGB;
-    export function hsl2rgb(h:number,s:number,l:number):RGB;
-    export function rgb2hsb(r:number,g:number,b:number):HSB;
-    export function rgb2hsl(r:number,g:number,b:number):HSL;
-
-    export function angle(x1:number,y1:number,x2:number,y2:number,x3?:number,y3?:number):number;
-    export function rad(deg:number):number;
-    export function deg(rad:number):number;
-    export function sin(angle: number): number;
-    export function cos(angle: number): number;
-    export function tan(angle: number): number;
-    export function asin(angle: number): number;
-    export function acos(angle: number): number;
-    export function atan(angle: number): number;
-    export function atan2(angle: number): number;
-
-    export function len(x1: number, y1: number, x2: number, y2: number): number;
-    export function len2(x1: number, y1: number, x2: number, y2: number): number;
-
-    export function parse(svg:string):Fragment;
-    export function parsePathString(pathString:string):Array<any>;
-    export function parsePathString(pathString:Array<string>):Array<any>;
-    export function parseTransformString(TString:string):Array<any>;
-    export function parseTransformString(TString:Array<string>):Array<any>;
-
-    export function closest(x: number, y: number, X: number, Y: number): boolean;
-
-    export interface RGB {
+
+declare module "snapsvg" {
```

Comment:
I don't think you should be using an ambient module declaration. I think `export as namespace` is indeed the more appropriate form here.

Reply:
Hi Daniel, I agree with you that a namespace looks cleaner than an ambient declaration. But the current TypeScript version does not support Declaration Merging when a namespace has the same name as a function out of the namespace.

Snap, by architecture, allows plugins that add certain functions to elements. My usecase is to add "translate", "rotate" and "scale" functions to the paper interface in JS. When taking that plugin scheme to TS, Declaration Merging is the appropriate way:

```namespace Snap {
    export interface Paper {
        translate(dx: Number, dy: Number): Snap.Paper;
        scale(sx: Number, sy: Number): Snap.Paper;
        rotate(angle: Number, cx: Number, cy: Number): Snap.Paper;
    }
    export interface Element {
        translate(dx: Number, dy: Number): Snap.Element;
        scale(sx: Number, sy: Number): Snap.Element;
        rotate(angle: Number, cx: Number, cy: Number): Snap.Element;
    }
}```

As long as the Snap(....) functions exist, overloading of both interfaces is not possible. That's the reason why /jquery/index.ts uses the same mechanics as an ambient module.

If you can provide a way, how Declaration Merging works with the current TSC version, I would be happy to use it and to cancel my pull request.

Reply:
Users can use module augmentation to reopen the interfaces: